File: mobile/android/core/auth/src/main/kotlin/com/slopssaloon/omen/core/auth/AuthFlow.kt
package com.slopssaloon.omen.core.auth

import com.slopssaloon.omen.core.session.Session

/**
 * Transient sign-in flow state (distinct from the persistent
 * [com.slopssaloon.omen.core.session.SessionState]). Every non-success state has a named,
 * safe next action so no screen shows a generic endless "Loading…" (M0a §6/§7, governance §5).
 *
 * The reducer ([AuthFlowReducer.reduce]) is pure — no coroutines, no Android, no network — so
 * the whole flow is exhaustively unit-testable. Side effects (calling [AuthRepository],
 * launching Credential Manager) are performed by the caller in response to the *entered* state.
 */
sealed interface AuthFlowState {
    /** Nothing in progress; the account/auth entry surface is shown. */
    data object Idle : AuthFlowState

    /** Requesting an OTP for [email]; caller invokes `requestEmailOtp`. */
    data class RequestingOtp(val email: String) : AuthFlowState

    /** OTP delivered; awaiting the user's 6-digit code for [email]. */
    data class AwaitingOtp(val email: String) : AuthFlowState

    /** Verifying the entered code; caller invokes `verifyEmailOtp`. */
    data class VerifyingOtp(val email: String) : AuthFlowState

    /** Credential Manager sheet in progress; caller invokes the Google provider. */
    data object LaunchingGoogle : AuthFlowState

    /** Exchanging a Google ID token for a session; caller invokes `signInWithGoogleIdToken`. */
    data object ExchangingGoogleToken : AuthFlowState

    /**
     * Provider-agnostic OAuth browser handoff in progress (M4-Auth-Providers-v1 §2.1). Caller
     * launches the Custom Tabs / `ASWebAuthenticationSession` for [providerId] (e.g. `"discord"`).
     */
    data class LaunchingOAuth(val providerId: String) : AuthFlowState

    /** Deep link returned; caller invokes `exchangeOAuthCode(providerId, code, verifier)`. */
    data class ExchangingOAuthCode(val providerId: String) : AuthFlowState

    /** Platform passkey UI is presenting; caller invokes the [PasskeyProvider]. */
    data object LaunchingPasskey : AuthFlowState

    /** Sending the platform assertion to Supabase for verification. */
    data object ExchangingPasskeyAssertion : AuthFlowState

    /** Terminal success — caller persists [session] via the session manager. */
    data class Authenticated(val session: Session) : AuthFlowState

    /** A recoverable stop with a named [reason]; the UI offers the matching next action. */
    data class Failed(val reason: AuthFailure) : AuthFlowState
}

/** Named, log-safe failure reasons that each map to a concrete recovery affordance. */
enum class AuthFailure {
    INVALID_EMAIL,                 // → correct the email
    INVALID_CODE,                  // → re-enter or resend the code
    CANCELED,                      // → return to Idle; not a hard error
    NETWORK,                       // → retry
    TIMEOUT,                       // → retry
    SERVER,                        // → retry / try another method
    GOOGLE_UNAVAILABLE,            // → use email instead
    NEEDS_REAUTH,                  // → sign in again
    OAUTH_PROVIDER_NOT_CONFIGURED, // → hide the button / use another provider
    OAUTH_CALLBACK_MISMATCH,       // → restart the OAuth flow (CSRF)
    PASSKEY_UNAVAILABLE,           // → use another method on this device
    PASSKEY_NO_CREDENTIAL,         // → sign in with another method, then pair a passkey
    UNKNOWN,                       // → retry / support without credentials
}

sealed interface AuthEvent {
    data class EmailSubmitted(val email: String) : AuthEvent
    data class OtpRequestResult(val outcome: AuthOutcome) : AuthEvent
    data class OtpSubmitted(val code: String) : AuthEvent
    data class OtpVerifyResult(val outcome: AuthOutcome) : AuthEvent
    data object GoogleRequested : AuthEvent
    data class GoogleTokenResult(val result: GoogleIdTokenResult) : AuthEvent
    data class GoogleExchangeResult(val outcome: AuthOutcome) : AuthEvent

    // M4-Auth-Providers-v1 §2.2 — provider-agnostic OAuth (e.g. Discord)
    data class OAuthRequested(val providerId: String) : AuthEvent
    data class OAuthCallbackReceived(
        val providerId: String,
        val code: String,
        val state: String,
    ) : AuthEvent
    data class OAuthExchangeResult(val providerId: String, val outcome: AuthOutcome) : AuthEvent

    // M4-Auth-Providers-v1 §2.2 — WebAuthn passkeys (different technology from OAuth)
    data object PasskeyRequested : AuthEvent
    data class PasskeyAssertionResult(val result: PasskeyResult) : AuthEvent
    data class PasskeyExchangeResult(val outcome: AuthOutcome) : AuthEvent

    data object Canceled : AuthEvent
    data object Reset : AuthEvent
}

object AuthFlowReducer {

    fun reduce(state: AuthFlowState, event: AuthEvent): AuthFlowState = when (event) {
        is AuthEvent.Reset -> AuthFlowState.Idle
        is AuthEvent.Canceled -> AuthFlowState.Failed(AuthFailure.CANCELED)

        is AuthEvent.EmailSubmitted ->
            if (EmailValidator.isValid(event.email)) AuthFlowState.RequestingOtp(EmailValidator.normalize(event.email))
            else AuthFlowState.Failed(AuthFailure.INVALID_EMAIL)

        is AuthEvent.OtpRequestResult -> when (val o = event.outcome) {
            is AuthOutcome.OtpSent -> (state as? AuthFlowState.RequestingOtp)
                ?.let { AuthFlowState.AwaitingOtp(it.email) } ?: AuthFlowState.Failed(AuthFailure.UNKNOWN)
            is AuthOutcome.RetryableError -> AuthFlowState.Failed(o.code.toFailure())
            else -> AuthFlowState.Failed(AuthFailure.UNKNOWN)
        }

        is AuthEvent.OtpSubmitted -> when (state) {
            is AuthFlowState.AwaitingOtp ->
                if (OtpCodeValidator.isValid(event.code)) AuthFlowState.VerifyingOtp(state.email)
                else AuthFlowState.Failed(AuthFailure.INVALID_CODE)
            else -> AuthFlowState.Failed(AuthFailure.UNKNOWN)
        }

        is AuthEvent.OtpVerifyResult -> when (val o = event.outcome) {
            is AuthOutcome.Success -> AuthFlowState.Authenticated(o.session)
            is AuthOutcome.InvalidCode -> AuthFlowState.Failed(AuthFailure.INVALID_CODE)
            is AuthOutcome.RetryableError -> AuthFlowState.Failed(o.code.toFailure())
            is AuthOutcome.NeedsReauth -> AuthFlowState.Failed(AuthFailure.NEEDS_REAUTH)
            else -> AuthFlowState.Failed(AuthFailure.UNKNOWN)
        }

        is AuthEvent.GoogleRequested -> AuthFlowState.LaunchingGoogle

        is AuthEvent.GoogleTokenResult -> when (event.result) {
            is GoogleIdTokenResult.Token -> AuthFlowState.ExchangingGoogleToken
            is GoogleIdTokenResult.Canceled -> AuthFlowState.Failed(AuthFailure.CANCELED)
            is GoogleIdTokenResult.Unavailable -> AuthFlowState.Failed(AuthFailure.GOOGLE_UNAVAILABLE)
            is GoogleIdTokenResult.Failed -> AuthFlowState.Failed(AuthFailure.UNKNOWN)
        }

        is AuthEvent.GoogleExchangeResult -> when (val o = event.outcome) {
            is AuthOutcome.Success -> AuthFlowState.Authenticated(o.session)
            is AuthOutcome.Unsupported -> AuthFlowState.Failed(AuthFailure.GOOGLE_UNAVAILABLE)
            is AuthOutcome.RetryableError -> AuthFlowState.Failed(o.code.toFailure())
            is AuthOutcome.NeedsReauth -> AuthFlowState.Failed(AuthFailure.NEEDS_REAUTH)
            else -> AuthFlowState.Failed(AuthFailure.UNKNOWN)
        }

        is AuthEvent.OAuthRequested -> AuthFlowState.LaunchingOAuth(event.providerId)

        is AuthEvent.OAuthCallbackReceived -> when (state) {
            is AuthFlowState.LaunchingOAuth ->
                if (state.providerId == event.providerId) AuthFlowState.ExchangingOAuthCode(event.providerId)
                // A stray callback that names a different providerId while we were launching
                // provider X is a routing/CSRF anomaly, not a normal recoverable event.
                else AuthFlowState.Failed(AuthFailure.OAUTH_CALLBACK_MISMATCH)
            else -> AuthFlowState.Failed(AuthFailure.UNKNOWN)
        }

        is AuthEvent.OAuthExchangeResult -> when (val o = event.outcome) {
            is AuthOutcome.Success -> AuthFlowState.Authenticated(o.session)
            is AuthOutcome.OAuthCallbackMismatch -> AuthFlowState.Failed(AuthFailure.OAUTH_CALLBACK_MISMATCH)
            is AuthOutcome.OAuthProviderNotConfigured -> AuthFlowState.Failed(AuthFailure.OAUTH_PROVIDER_NOT_CONFIGURED)
            is AuthOutcome.Canceled -> AuthFlowState.Failed(AuthFailure.CANCELED)
            is AuthOutcome.RetryableError -> AuthFlowState.Failed(o.code.toFailure())
            is AuthOutcome.NeedsReauth -> AuthFlowState.Failed(AuthFailure.NEEDS_REAUTH)
            else -> AuthFlowState.Failed(AuthFailure.UNKNOWN)
        }

        is AuthEvent.PasskeyRequested -> AuthFlowState.LaunchingPasskey

        is AuthEvent.PasskeyAssertionResult -> when (event.result) {
            is PasskeyResult.Assertion -> AuthFlowState.ExchangingPasskeyAssertion
            is PasskeyResult.Canceled -> AuthFlowState.Failed(AuthFailure.CANCELED)
            is PasskeyResult.Unavailable -> AuthFlowState.Failed(AuthFailure.PASSKEY_UNAVAILABLE)
            is PasskeyResult.NoCredential -> AuthFlowState.Failed(AuthFailure.PASSKEY_NO_CREDENTIAL)
            is PasskeyResult.Failed -> AuthFlowState.Failed(AuthFailure.UNKNOWN)
        }

        is AuthEvent.PasskeyExchangeResult -> when (val o = event.outcome) {
            is AuthOutcome.Success -> AuthFlowState.Authenticated(o.session)
            is AuthOutcome.RetryableError -> AuthFlowState.Failed(o.code.toFailure())
            is AuthOutcome.NeedsReauth -> AuthFlowState.Failed(AuthFailure.NEEDS_REAUTH)
            else -> AuthFlowState.Failed(AuthFailure.UNKNOWN)
        }
    }

    private fun RetryableCode.toFailure(): AuthFailure = asAuthFailure()
}

/**
 * Top-level so the OTP resend path can reuse it. That path surfaces its failure beside the code
 * field instead of reducing it into `AuthFlowState.Failed`, because a reduced failure knocks the
 * user out of `AwaitingOtp` and discards the code they may be mid-way through typing — but it
 * still needs the same mapping, and a second copy would be a second thing to keep in step.
 */
fun RetryableCode.asAuthFailure(): AuthFailure = when (this) {
    RetryableCode.NETWORK -> AuthFailure.NETWORK
    RetryableCode.TIMEOUT -> AuthFailure.TIMEOUT
    RetryableCode.SERVER -> AuthFailure.SERVER
    RetryableCode.UNKNOWN -> AuthFailure.UNKNOWN
}

File: mobile/ios/OmenIOS/OmenIOS/Core/Auth/AuthFlow.swift
import Foundation

/// Mirrors Android `core/auth/AuthFlow.kt`'s `AuthFlowState`. Apple states replace Google states
/// (iOS's primary native mechanism is Apple, not Google) — same shape, different provider.
enum AuthFlowState: Equatable {
    case idle
    case requestingOtp(email: String)
    case awaitingOtp(email: String)
    case verifyingOtp(email: String)
    case launchingApple
    case exchangingAppleToken
    // M4-Auth-Providers-v1 §2.1 — provider-agnostic OAuth (e.g. Discord)
    case launchingOAuth(providerId: String)
    case exchangingOAuthCode(providerId: String)
    // M4-Auth-Providers-v1 §2.1 — WebAuthn passkeys (different technology from OAuth)
    case launchingPasskey
    case exchangingPasskeyAssertion
    case authenticated(session: Session)
    case failed(reason: AuthFailure)
}

/// Mirrors Android `core/auth/AuthFlow.kt`'s `AuthFailure`. Each case has a documented recovery
/// affordance in `AuthFailure.userMessage` — never a raw provider error string (M0c §8).
enum AuthFailure: Equatable {
    case invalidEmail
    case invalidCode
    case canceled
    case network
    case timeout
    case server
    case appleUnavailable
    case needsReauth
    // M4-Auth-Providers-v1 §2.3
    case oauthProviderNotConfigured
    case oauthCallbackMismatch
    case passkeyUnavailable
    case passkeyNoCredential
    case unknown
}

extension AuthFailure {
    /// Safe, opaque, user-facing copy. Never derived from a raw HTTP status or provider message.
    var userMessage: String {
        switch self {
        case .invalidEmail: return "Enter a valid email address."
        case .invalidCode: return "That code didn't match. Check your email and try again."
        case .canceled: return "Sign-in was canceled."
        case .network: return "No connection. Check your network and try again."
        case .timeout: return "That took too long. Try again."
        case .server: return "Omen is having trouble right now. Try again shortly."
        case .appleUnavailable: return "Sign in with Apple isn't available right now. Use email instead."
        case .needsReauth: return "Sign in again to continue."
        // M4-Auth-Providers-v1: placeholder copy per brief §10; slops-ux-copy tunes wording
        // after the mechanical seam works. Keeping strings functional so nothing is blank.
        case .oauthProviderNotConfigured: return "That sign-in option isn't available right now. Use another method."
        case .oauthCallbackMismatch: return "Sign-in couldn't be verified. Start again."
        case .passkeyUnavailable: return "Passkeys aren't available on this device. Use another method."
        case .passkeyNoCredential: return "No passkey found for this account on this device. Sign in another way to pair one."
        case .unknown: return "Something went wrong. Try again."
        }
    }
}

/// Mirrors Android `core/auth/AuthFlow.kt`'s `AuthEvent`.
enum AuthEvent {
    case emailSubmitted(email: String)
    case otpRequestResult(outcome: AuthOutcome)
    case otpSubmitted(code: String)
    case otpVerifyResult(outcome: AuthOutcome)
    case appleRequested
    case appleTokenResult(AppleIDTokenResult)
    case appleExchangeResult(outcome: AuthOutcome)
    // M4-Auth-Providers-v1 §2.2 — provider-agnostic OAuth (e.g. Discord)
    case oauthRequested(providerId: String)
    case oauthCallbackReceived(providerId: String, code: String, state: String)
    case oauthExchangeResult(providerId: String, outcome: AuthOutcome)
    // M4-Auth-Providers-v1 §2.2 — WebAuthn passkeys
    case passkeyRequested
    case passkeyAssertionResult(PasskeyResult)
    case passkeyExchangeResult(outcome: AuthOutcome)
    case canceled
    case reset
}

/// Mirrors Android `core/auth/AuthFlow.kt`'s `AuthFlowReducer`: a pure, exhaustive, side-effect
/// free state machine. No coroutines/async, no Foundation networking — just data in, data out —
/// so it is trivially unit-testable (see `AuthFlowReducerTests`).
enum AuthFlowReducer {
    static func reduce(state: AuthFlowState, event: AuthEvent) -> AuthFlowState {
        switch event {
        case .reset:
            return .idle

        case .canceled:
            return .failed(reason: .canceled)

        case .emailSubmitted(let email):
            guard EmailValidator.isValid(email) else {
                return .failed(reason: .invalidEmail)
            }
            return .requestingOtp(email: EmailValidator.normalize(email))

        case .otpRequestResult(let outcome):
            guard case .requestingOtp(let email) = state else { return state }
            return reduceOtpRequestResult(outcome, email: email)

        case .otpSubmitted(let code):
            guard case .awaitingOtp(let email) = state else { return state }
            guard OtpCodeValidator.isValid(code) else {
                return .failed(reason: .invalidCode)
            }
            return .verifyingOtp(email: email)

        case .otpVerifyResult(let outcome):
            guard case .verifyingOtp = state else { return state }
            return reduceCredentialOutcome(outcome, invalidCodeFailure: .invalidCode)

        case .appleRequested:
            return .launchingApple

        case .appleTokenResult(let result):
            guard case .launchingApple = state else { return state }
            switch result {
            case .token:
                return .exchangingAppleToken
            case .canceled:
                return .failed(reason: .canceled)
            case .unavailable:
                return .failed(reason: .appleUnavailable)
            case .failed:
                return .failed(reason: .unknown)
            }

        case .appleExchangeResult(let outcome):
            guard case .exchangingAppleToken = state else { return state }
            return reduceCredentialOutcome(outcome, invalidCodeFailure: .unknown)

        case .oauthRequested(let providerId):
            return .launchingOAuth(providerId: providerId)

        case .oauthCallbackReceived(let providerId, _, _):
            guard case .launchingOAuth(let launchingProviderId) = state else {
                return .failed(reason: .unknown)
            }
            guard launchingProviderId == providerId else {
                // Stray callback naming a different providerId while launching provider X is a
                // routing/CSRF anomaly, not a normal recoverable event.
                return .failed(reason: .oauthCallbackMismatch)
            }
            return .exchangingOAuthCode(providerId: providerId)

        case .oauthExchangeResult(_, let outcome):
            return reduceOAuthExchangeOutcome(outcome)

        case .passkeyRequested:
            return .launchingPasskey

        case .passkeyAssertionResult(let result):
            guard case .launchingPasskey = state else { return state }
            switch result {
            case .assertion:
                return .exchangingPasskeyAssertion
            case .canceled:
                return .failed(reason: .canceled)
            case .unavailable:
                return .failed(reason: .passkeyUnavailable)
            case .noCredential:
                return .failed(reason: .passkeyNoCredential)
            case .failed:
                return .failed(reason: .unknown)
            }

        case .passkeyExchangeResult(let outcome):
            return reduceCredentialOutcome(outcome, invalidCodeFailure: .unknown)
        }
    }

    private static func reduceOAuthExchangeOutcome(_ outcome: AuthOutcome) -> AuthFlowState {
        switch outcome {
        case .success(let session):
            return .authenticated(session: session)
        case .oauthCallbackMismatch:
            return .failed(reason: .oauthCallbackMismatch)
        case .oauthProviderNotConfigured:
            return .failed(reason: .oauthProviderNotConfigured)
        case .canceled:
            return .failed(reason: .canceled)
        case .retryableError(let code):
            return .failed(reason: code.asAuthFailure)
        case .needsReauth:
            return .failed(reason: .needsReauth)
        case .otpSent, .invalidCode, .unsupported:
            return .failed(reason: .unknown)
        }
    }

    private static func reduceOtpRequestResult(_ outcome: AuthOutcome, email: String) -> AuthFlowState {
        switch outcome {
        case .otpSent, .success:
            return .awaitingOtp(email: email)
        case .canceled:
            return .failed(reason: .canceled)
        case .retryableError(let code):
            return .failed(reason: code.asAuthFailure)
        case .invalidCode, .needsReauth, .unsupported,
             .oauthCallbackMismatch, .oauthProviderNotConfigured:
            return .failed(reason: .unknown)
        }
    }

    /// Shared by OTP verification, Apple token exchange, and passkey assertion exchange: all
    /// resolve a credential against the server and either land on `.authenticated` or a named
    /// `.failed` reason.
    private static func reduceCredentialOutcome(_ outcome: AuthOutcome, invalidCodeFailure: AuthFailure) -> AuthFlowState {
        switch outcome {
        case .success(let session):
            return .authenticated(session: session)
        case .invalidCode:
            return .failed(reason: invalidCodeFailure)
        case .canceled:
            return .failed(reason: .canceled)
        case .retryableError(let code):
            return .failed(reason: code.asAuthFailure)
        case .needsReauth:
            return .failed(reason: .needsReauth)
        case .otpSent, .unsupported,
             .oauthCallbackMismatch, .oauthProviderNotConfigured:
            return .failed(reason: .unknown)
        }
    }
}

/// Internal rather than fileprivate: the resend path in `AuthViewModel` surfaces its failure
/// beside the code field instead of through the reducer, so that a failed resend cannot knock
/// the user out of code entry and discard what they have typed. It still needs the same
/// mapping, and a second copy of it would be a second thing to keep in step.
extension RetryableCode {
    var asAuthFailure: AuthFailure {
        switch self {
        case .network: return .network
        case .timeout: return .timeout
        case .server: return .server
        case .unknown: return .unknown
        }
    }
}
